integration/ct_integration.go

```diff
@@ -0,0 +1,636 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"bytes"
+	"context"
+	basecrypto "crypto"
+	cryptorand "crypto/rand"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	ct "github.com/google/certificate-transparency/go"
+	"github.com/google/certificate-transparency/go/client"
+	"github.com/google/certificate-transparency/go/jsonclient"
+	"github.com/google/certificate-transparency/go/merkletree"
+	"github.com/google/certificate-transparency/go/tls"
+	"github.com/google/certificate-transparency/go/x509"
+	"github.com/google/certificate-transparency/go/x509/pkix"
+	"github.com/google/trillian/crypto"
+	ctfe "github.com/google/trillian/examples/ct"
+	"github.com/google/trillian/testonly"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+// TODO(drysdale): convert to use trillian/merkle to avoid the dependency on the
+// CT code (which in turn requires C++ code).
+
+// Verifier is used to verify Merkle tree calculations.
+var Verifier = merkletree.NewMerkleVerifier(func(data []byte) []byte {
+	hash := sha256.Sum256(data)
+	return hash[:]
+})
+
+// ClientPool holds a collection of CT LogClient instances.
+type ClientPool []*client.LogClient
+
+// Pick a random client from the pool.
+func (p ClientPool) Pick() *client.LogClient {
+	if len(p) == 0 {
+		return nil
+	}
+	return p[rand.Intn(len(p))]
+}
+
+// RunCTIntegrationForLog tests against the log with configuration cfg, with a set
+// of comma-separated server addresses given by servers, assuming that testdir holds
+// a variety of test data files.
+func RunCTIntegrationForLog(cfg ctfe.LogConfig, servers, testdir string, stats *wantStats) error {
+	opts := jsonclient.Options{}
+	if cfg.PubKeyPEMFile != "" {
+		pubkey, err := ioutil.ReadFile(cfg.PubKeyPEMFile)
+		if err != nil {
+			return fmt.Errorf("failed to get public key contents: %v", err)
```

Comment:
Log some identifier so that errors can be tied to the specific log/config? (elsewhere too)